Add support for Error objects in toMatchObject

jseminck opened the pull request:
**Summary**
https://github.com/facebook/jest/issues/3859 - Add support for `Error` objects in `toMatchObject()`

I didn't realise the original PR got closed after my suggestion, and then I completely lost track of it. Sorry for that: https://github.com/facebook/jest/pull/3866

Explanation of this fix: https://github.com/facebook/jest/pull/3866#issuecomment-313605201

I also moved some functions to the `utils.js` file.

**Test plan**
Unit tests were added

cpojer commented on integration_tests/toMatchSnapshot/__tests__/basic-support.test.js:
```diff
@@ -0,0 +1 @@
+test('snapshots', () => expect({apple: "updated value"}).toMatchSnapshot());
```
This needs Jest's Copyright header, and a trailing newline please.

cpojer replied:
Actually, I don't understand the purpose of this test. Why is it here?

jseminck replied:
I think my tests went badly. I was running on Windows and a bunch of stuff was failing. Will fix now that I'm back on the mac.

jseminck replied:
Aha, I believe these files got added because I interrupted a test run, and it creates/deletes files during tests but when you interrupt it doesn't clean up properly. Mystery solved 👌

cpojer commented:
Please add type annotations and make sure `yarn test` passes locally.

jseminck commented:
Hi, thanks for the reply. I fixed the files that got added by accident.

Regarding tests: The tests on master do not pass for me, nor on my mac or on my windows machine. I checked that the tests which I added pass (`matchers.test.js`).

I'll look into adding the type definitions as well, but I don't really understand why it is needed as I didn't change any of the function signatures. Is it related to moving the functions to the `utils` file?

cpojer commented:
Thank you!
